### PR TITLE
SCC updates needed for ecWam

### PR DESCRIPTION
**Pull request by awnawab**
For both correctness and performance, the SCC transform needed a few updates before it could be applied to the ecWam physics. With that aim, this branch adds the following features:
- `trim_vector_sections` option to remove scalar assignments from vector loops
- Removal of calls explicitly marked with `!$loki inline` pragmas from separator nodes
- Loki-pragma driven insertion of reduction loops

The companion piece to this PR is the following ecWam branch where similar to CLOUDSC, we use Loki's cmake tools to build a range of transformations: https://github.com/ecmwf-ifs/ecwam/tree/naan-loki-trafo

NB: This is stacked on top of #92.

### CODE REVIEW

File: transformations/transformations/pool_allocator.py
Comment (mlange05) [inline transformations/transformations/pool_allocator.py:583]: [no action] Is this just to catch optional whitespace? Just for my own understanding, why is this an issue?

**Review by mlange05 — CHANGES_REQUESTED** [review]
Ok, overall this looks really great, and contains several very useful things that I will need to try out quickly on my code sections! Some of these could've probably been standalone PRs, but that's not super-important here, as the PR is still neatly sized.

I have one comment though, that I think needs addressing before we can land this. The use of `!$loki inline` suggests way more than we do atm, and while we might want to fully support explicit full inlining macros later, I'd prefer if we either rename it, or look at the style of routine to determine the override that this effectively brings.

All other comments are optional or just that, "comments", so GTG once the pragma is addressed. I'll run EC-phys regression next before merging and post back here, but should be fine.

File: transformations/tests/test_single_column_coalesced.py
Comment (awnawab) [reply transformations/tests/test_single_column_coalesced.py:1547]: Yes I think it should be possible to do this by examining whether we have any horizontal loops inside the attached routine. I'll have a go at that, thanks for the suggestion!

File: transformations/tests/test_single_column_coalesced.py
Comment (awnawab) [reply transformations/tests/test_single_column_coalesced.py:1721]: No it doesn't work, I get an `argument of type 'Section' is not iterable` error.

File: transformations/transformations/pool_allocator.py
Comment (awnawab) [reply transformations/transformations/pool_allocator.py:583]: If the pragma had both a `private` and a `firstprivate` clause, then the previous implementation would put `YLSTACK` in both.

**Comment by awnawab** [comment]
Thanks a lot for the review @mlange05! Whilst I address the `loki inline` pragma comment, I also wanted to just highlight that with these changes the modular scc transform now diverges from the old scc, which now lives in `single_column_coalesced_wrapper.py`. Please let me know if you would like me to play these changes through the scc wrapper too.

**Comment by mlange05** [comment]
> Thanks a lot for the review @mlange05! Whilst I address the `loki inline` pragma comment, I also wanted to just highlight that with these changes the modular scc transform now diverges from the old scc, which now lives in `single_column_coalesced_wrapper.py`. Please let me know if you would like me to play these changes through the scc wrapper too.

No, that's fine. In fact it's a good motivator for me to flip the `ecphys` mode over to use the same a convert. No need to hold up this PR for this.

File: transformations/tests/test_single_column_coalesced.py
Comment (mlange05) [reply transformations/tests/test_single_column_coalesced.py:1722]: Oh, ok. We'll need to fix this at some point (famous last words :grin: ). Was just out of interest, nothing needed for this PR.

File: transformations/transformations/pool_allocator.py
Comment (mlange05) [reply transformations/transformations/pool_allocator.py:583]: Ahhhhh! Makes sense, thanks.

File: transformations/transformations/single_column_coalesced_vector.py
Comment (mlange05) [inline transformations/transformations/single_column_coalesced_vector.py:82]: Uhhh, that's nifty, but could also be quite error-prone. If I understand correctly, if the callee routine deviates from the coding norms, it will now be treated as effectively sequential/non-vector? That could be dangerous as that actually happens quite often in unsanitised code. 

I had originally thought, looking explicitly for the `!$loki routine seq` pragma on the callee (which we already use to bypass SCC-vectorisation), would be more appropriate here (although not quite the one-liner that this is, admittedly)? Would be keen to hear counter-points if I'm not seeing something here? @awnawab

Comment (awnawab) [reply transformations/transformations/single_column_coalesced_vector.py:82]: Ah yes, fully agreed, if we already have one mechanism to mark routines as sequential, then we should just use that. Thanks for the suggestion, I'll implement that! And also add more tests to boost coverage 😆

Comment (mlange05) [reply transformations/transformations/single_column_coalesced_vector.py:82]: Also, paging Dr @reuterbal for input on this one?

Comment (reuterbal) [reply transformations/transformations/single_column_coalesced_vector.py]: Yeah, I might have to come back to that one as I do have quite a few routines that are unsanitised in that sense. I agree that checking only for horizontal size may be too error-prone, though, so sticking to the existing annotation mechanism would probably be better.

File: transformations/tests/test_utility_routines.py
Comment (reuterbal) [reply transformations/tests/test_utility_routines.py:96]: No `use rick_astley`, though?

**Review by reuterbal — APPROVED** [review]
Thanks a lot, looks good to me!

File: scripts/loki_transform.py
Comment (awnawab) [reply scripts/loki_transform.py:301]: Yes! file graph makes so much more sense, I no longer have to see unintelligibly long log entries from the `FileWriteTransformation`. Thanks! 😄

**Pull request merged**
